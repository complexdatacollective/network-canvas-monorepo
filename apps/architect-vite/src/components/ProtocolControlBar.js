import { Button, Spinner } from "@codaco/ui";
import { useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import ControlBar from "~/src/components/ControlBar";
import { UnsavedChanges } from "~/src/components/Dialogs";
import { actionCreators as dialogActions } from "~/src/ducks/modules/dialogs";
import { actionCreators as sessionActions } from "~/src/ducks/modules/session";
import { selectors as statusSelectors } from "~/src/ducks/modules/ui/status";
import { actionLocks as protocolsLocks, actionCreators as userActions } from "~/src/ducks/modules/userActions";
import logoutIcon from "~/src/images/home/log-out.svg";
import { getHasUnsavedChanges, getIsProtocolValid } from "~/src/selectors/session";

const unsavedChangesDialog = UnsavedChanges({
	message: <p>Your protocol has changes that have not yet been saved. Continuing will discard these changes!</p>,
	confirmLabel: "Discard Changes",
});

const ProtocolControlBar = () => {
	const dispatch = useDispatch();

	const hasUnsavedChanges = useSelector((state) => getHasUnsavedChanges(state));
	const isSaving = useSelector((state) => statusSelectors.getIsBusy(state, protocolsLocks.saving));
	const protocolIsValid = useSelector((state) => getIsProtocolValid(state));
	const saveNetcanvas = useCallback(() => dispatch(userActions.saveNetcanvas()), [dispatch]);

	const handleClickStart = useCallback(
		() =>
			Promise.resolve()
				.then(() => {
					if (!hasUnsavedChanges) {
						return true;
					}

					return dispatch(dialogActions.openDialog(unsavedChangesDialog));
				})
				.then((confirm) => {
					if (!confirm) {
						return;
					}
					dispatch(sessionActions.resetSession());
				}),
		[dispatch, hasUnsavedChanges],
	);

	return (
		<ControlBar
			show
			secondaryButtons={[
				<Button
					key="return-button"
					color="platinum"
					icon={
						<div>
							<img src={logoutIcon} alt="Return to start screen" />
						</div>
					}
					onClick={handleClickStart}
				>
					Return to start screen
				</Button>,
			]}
			buttons={[
				...(protocolIsValid && hasUnsavedChanges
					? [
							<Button
								key="save-button"
								onClick={saveNetcanvas}
								color="primary"
								data-variant="save"
								disabled={isSaving}
								content={isSaving ? "Saving..." : "Save Changes"}
								iconPosition="right"
								icon={
									isSaving ? (
										<div>
											<Spinner size="0.5rem" />
										</div>
									) : (
										"arrow-right"
									)
								}
							/>,
						]
					: []),
			]}
		/>
	);
};

export default ProtocolControlBar;
