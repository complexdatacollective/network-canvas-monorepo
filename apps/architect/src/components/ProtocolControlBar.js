import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { UnsavedChanges } from '@components/Dialogs';
import { Button, Spinner } from '@codaco/ui';
import { getHasUnsavedChanges, getIsProtocolValid } from '@selectors/session';
import { actionCreators as dialogActions } from '@modules/dialogs';
import { actionCreators as userActions, actionLocks as protocolsLocks } from '@modules/userActions';
import { actionCreators as sessionActions } from '@modules/session';
import { selectors as statusSelectors } from '@modules/ui/status';
import logoutIcon from '@app/images/home/log-out.svg';
import ControlBar from '@components/ControlBar';

const unsavedChangesDialog = UnsavedChanges({
  message: (
    <p>
      Your protocol has changes that have not yet been saved. Continuing will
      discard these changes!
    </p>
  ),
  confirmLabel: 'Discard Changes',
});

const ProtocolControlBar = () => {
  const dispatch = useDispatch();

  const hasUnsavedChanges = useSelector((state) => getHasUnsavedChanges(state));
  const isSaving = useSelector((state) => statusSelectors.getIsBusy(state, protocolsLocks.saving));
  const protocolIsValid = useSelector((state) => getIsProtocolValid(state));
  const saveNetcanvas = useCallback(() => dispatch(userActions.saveNetcanvas()), [dispatch]);

  const handleClickStart = useCallback(
    () => Promise.resolve()
      .then(() => {
        if (!hasUnsavedChanges) { return true; }

        return dispatch(dialogActions.openDialog(unsavedChangesDialog));
      })
      .then((confirm) => {
        if (!confirm) { return; }
        dispatch(sessionActions.resetSession());
      }),
    [dispatch, hasUnsavedChanges],
  );

  return (
    <ControlBar
      show
      secondaryButtons={[
        <Button
          key="return-button"
          color="platinum"
          icon={<div><img src={logoutIcon} alt="Return to start screen" /></div>}
          onClick={handleClickStart}
        >
          Return to start screen
        </Button>,
      ]}
      buttons={[
        ...(protocolIsValid && hasUnsavedChanges ? [<Button
          key="save-button"
          onClick={saveNetcanvas}
          color="primary"
          data-variant="save"
          disabled={isSaving}
          content={isSaving ? 'Saving...' : 'Save Changes'}
          iconPosition="right"
          icon={isSaving ? <div><Spinner size="0.5rem" /></div> : 'arrow-right'}
        />] : []),
      ]}
    />
  );
};

export default ProtocolControlBar;
