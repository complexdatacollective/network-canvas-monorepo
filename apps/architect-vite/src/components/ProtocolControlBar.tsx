import { Check, Download } from "lucide-react";
import { useCallback, useState } from "react";
import { useLocation } from "wouter";
import ControlBar from "~/components/ControlBar";
import { useAppDispatch } from "~/ducks/hooks";
import { clearActiveProtocol } from "~/ducks/modules/activeProtocol";
import { actionCreators as dialogActions } from "~/ducks/modules/dialogs";
import { checkUnsavedChanges, exportProtocol } from "~/ducks/modules/userActions/userActions";
import logoutIcon from "~/images/home/log-out.svg";
import { Button } from "~/lib/legacy-ui/components";

const ProtocolControlBar = () => {
	const dispatch = useAppDispatch();
	const [, navigate] = useLocation();
	const [isExporting, setIsExporting] = useState(false);
	const [downloadSuccess, setDownloadSuccess] = useState(false);

	const handleReturnToStart = useCallback(async () => {
		// todo: confirm un-downloaded changes

		const canContinue = await dispatch(checkUnsavedChanges()).unwrap();
		if (!canContinue) {
			return;
		}

		dispatch(
			dialogActions.openDialog({
				type: "Warning",
				title: "Clear Editor?",
				message:
					"Returning to the start screen will clear the current protocol from the editor. If you have made changes to your protocol, please ensure you have downloaded the updated version before proceeding.",
				confirmLabel: "Return to start screen",
				onConfirm: async () => {
					await dispatch(clearActiveProtocol());
					navigate("/");
				},
			}) as any,
		);
	}, [dispatch, navigate]);

	const handleDownload = useCallback(async () => {
		try {
			setIsExporting(true);
			await dispatch(exportProtocol()).unwrap();
			setDownloadSuccess(true);
			setTimeout(() => setDownloadSuccess(false), 2000);
		} catch (error) {
			throw new Error(`Failed to export protocol: ${error}`);
		} finally {
			setIsExporting(false);
		}
	}, [dispatch]);

	return (
		<ControlBar
			secondaryButtons={[
				<Button
					key="return-button"
					color="platinum"
					icon={
						<div>
							<img src={logoutIcon} alt="Return to start screen" />
						</div>
					}
					onClick={handleReturnToStart}
				>
					Return to start screen
				</Button>,
			]}
			buttons={[
				<Button
					key="export-button"
					onClick={handleDownload}
					color="sea-green"
					content={downloadSuccess ? "Downloaded" : isExporting ? "Downloading..." : "Download"}
					disabled={isExporting}
					icon={downloadSuccess ? <Check size={22} /> : <Download size={22} />}
				/>,
			]}
		/>
	);
};

export default ProtocolControlBar;
