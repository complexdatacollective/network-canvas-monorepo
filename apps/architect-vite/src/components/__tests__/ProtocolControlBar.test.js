/* eslint-env jest */

import React from 'react';
import { mount } from 'enzyme';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import ProtocolControlBar from '../ProtocolControlBar';
import testState from '../../__tests__/testState.json';

const mockProps = {
  saveProtocol: () => {},
  isSaving: false,
  isDisabled: false,
  showControlBar: true,
  hasUnsavedChanges: false,
  hasAnyStages: true,
  handleClickStart: () => {},
};

const mockStore = createStore(() => testState);

describe('<ProtocolControlBar />', () => {
  it('can render', () => {
    const component = mount(
      <Provider store={mockStore}>
        <ProtocolControlBar {...mockProps} />
      </Provider>,
    );

    expect(component.html()).toMatchSnapshot();
  });
});
