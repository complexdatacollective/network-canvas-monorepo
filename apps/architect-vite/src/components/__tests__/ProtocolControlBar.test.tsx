import { describe, it, expect } from "vitest";

import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import ProtocolControlBar from "../ProtocolControlBar";
import testState from "../../__tests__/testState.json";

const mockProps = {
	saveProtocol: () => {},
	isSaving: false,
	isDisabled: false,
	showControlBar: true,
	hasUnsavedChanges: false,
	hasAnyStages: true,
	handleClickStart: () => {},
};

const mockStore = createStore(() => testState);

describe("<ProtocolControlBar />", () => {
	it("can render", () => {
		const { container } = render(
			<Provider store={mockStore}>
				<ProtocolControlBar {...mockProps} />
			</Provider>,
		);

		expect(container).toMatchSnapshot();
	});
});
