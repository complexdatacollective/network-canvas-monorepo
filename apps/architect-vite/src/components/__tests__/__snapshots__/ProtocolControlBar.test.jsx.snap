// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ProtocolControlBar /> > can render 1`] = `
<div>
  <div
    class="control-bar"
  >
    <div
      class="control-bar__secondary-buttons"
    >
      <div>
        <button
          class="button button--platinum button--has-icon"
          type="button"
        >
          <div
            class="button__icon"
          >
            <img
              alt="Return to start screen"
              src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20xmlns:xlink='http://www.w3.org/1999/xlink'%20height='24'%20width='24'%20viewBox='0%200%2024%2024'%3e%3cpath%20fill-rule='evenodd'%20fill='%236d6f76'%20clip-rule='evenodd'%20d='M22,18c0,1.1-0.9,2-2,2h-9c-1.1,0-2-0.9-2-2v-2c0-0.6,0.4-1,1-1c0.6,0,1,0.4,1,1v2h9V6h-9v2c0,0.6-0.4,1-1,1%20C9.4,9,9,8.6,9,8V6c0-1.1,0.9-2,2-2h9c1.1,0,2,0.9,2,2V18z%20M6.7,15.7c-0.4,0.4-1,0.4-1.4,0l-3-3c-0.4-0.4-0.4-1,0-1.4l3-3%20c0.4-0.4,1-0.4,1.4,0c0.4,0.4,0.4,1,0,1.4L5.4,11H15c0.6,0,1,0.4,1,1s-0.4,1-1,1H5.4l1.3,1.3C7.1,14.7,7.1,15.3,6.7,15.7z'/%3e%3c/svg%3e"
            />
          </div>
          <span
            class="button__content"
          >
            Return to start screen
          </span>
        </button>
      </div>
    </div>
    <div
      class="control-bar__primary-buttons"
    />
  </div>
</div>
`;
